aptos-move/block-executor/src/scheduler_v2.rs

```diff
@@ -4,21 +4,360 @@
+/// worker thread responsible for the owner transaction to:
+///
+/// 1.  Identify such dependent transactions.
+/// 2.  Attempt to initiate their abort by calling `[SchedulerV2::start_abort]` on the scheduler.
+/// 3.  Track the outcome of these abort attempts within its `invalidations` map.
+///
+/// It uses interior mutability (`RefCell`) for the `invalidations` map to manage this state.
+/// The `AbortManager` instance is provided by value to `[SchedulerV2::finish_execution]` to enforce
+/// a clear ownership pattern and ensure its state is consumed correctly after the owner
+/// transaction's execution attempt is processed by the scheduler.
+pub(crate) struct AbortManager<'a> {
+    owner_txn_idx: TxnIndex,
+    owner_incarnation: Incarnation,
+    scheduler: &'a SchedulerV2,
+    // Transaction index in the map implies a write by (owner_txn_idx, owner_incarnation)
+    // invalidated a read by the said transaction. If the incarnation is stored in the
+    // entry, then start_abort call was successful, implying a promise to call finish_abort.
+    invalidations: RefCell<BTreeMap<TxnIndex, Option<Incarnation>>>,
+}
+
+impl<'a> AbortManager<'a> {
+    pub(crate) fn new(
+        owner_txn_idx: TxnIndex,
+        owner_incarnation: Incarnation,
+        scheduler: &'a SchedulerV2,
+    ) -> Self {
+        Self {
+            owner_txn_idx,
+            owner_incarnation,
+            scheduler,
+            invalidations: RefCell::new(BTreeMap::new()),
+        }
+    }
+
+    pub(crate) fn invalidate_dependencies(
+        &self,
+        dependencies: BTreeSet<(TxnIndex, Incarnation)>,
+    ) -> Result<(), PanicError> {
+        for (txn_idx, incarnation) in dependencies {
+            self.invalidate(txn_idx, incarnation)?;
+        }
+        Ok(())
+    }
+
+    fn invalidate(
+        &self,
+        invalidated_txn_idx: TxnIndex,
+        invalidated_incarnation: Incarnation,
+    ) -> Result<(), PanicError> {
+        if invalidated_txn_idx <= self.owner_txn_idx {
+            return Err(code_invariant_error(format!(
+                "Execution of version ({}, {}) may not invalidate lower version ({}, {})",
+                self.owner_txn_idx,
+                self.owner_incarnation,
+                invalidated_txn_idx,
+                invalidated_incarnation,
+            )));
+        }
+
+        let mut invalidations = self.invalidations.borrow_mut();
+        match invalidations.entry(invalidated_txn_idx) {
+            Vacant(vacant_entry) => {
+                // For vacant entries, we always need to try abort.
+                // The result of start_abort (None or Some(incarnation)) is stored.
+                let _ = vacant_entry
+                    .insert(self.start_abort(invalidated_txn_idx, invalidated_incarnation)?);
+            },
+            Occupied(mut occupied_entry) => {
+                match occupied_entry.get() {
+                    None => {
```

Comment:
You dont need the none here

Reply:
Actually I remember why this is here - we need none because it allows us to build the dependency graph regardless of whether we succeeded in start abort or not.

Reply:
check finish_execution

Reply:
Ignore. 
None is used for the stalling mechanism in finish_execution().